Fix panic in attachDetachController

mshaverdo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
 /kind bug


**What this PR does / why we need it**:
PR fixes panics in `operationGenerator.GenerateDetachVolumeFunc()` in rare case of `volumePluginMgr.FindAttachablePluginByName()` finds plugin by name, but this plugin isn't a `AttachableVolumePlugin`. See #71463 
Although it clarifies  to callers of  ` VolumePluginMgr.FindAttachablePluginByName()` what cases they are expected to handle.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #71463

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
Fixes possible panic during volume detach, if corresponding volume plugin became non-attachable
```

gnufied commented on pkg/volume/plugins.go:
```diff
@@ -797,15 +797,18 @@ func (pm *VolumePluginMgr) FindAttachablePluginBySpec(spec *Spec) (AttachableVol
+		return attachablePlugin, true, nil
 	}
-	return nil, nil
+	return nil, false, nil
```
I think we need not touch this method. Rather than relying on `false` return value from here, we should just rely on `nil` plugin check in affected code. This would keep the diff small and ensure consistency between different `FindXXXPlugin` functions.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -386,8 +392,13 @@ func (og *operationGenerator) GenerateDetachVolumeFunc(
+		if !ok {
+			if err == nil {
+				err = fmt.Errorf("plugin %s isn't a AttachableVolumePlugin", pluginName)
+			}
```
I think it is simpler to just check for `nil`  `attachableVolumePlugin` in the code rather than introducing the boolean return.

gnufied commented:
/ok-to-test

mshaverdo commented:
@gnufied @saad-ali 

> I think it is simpler to just check for nil attachableVolumePlugin in the code rather than introducing the boolean return.

Yes, it's simpler, lead to minimal diff and completely fix this bug.
But clients of `FindXXXXXXPluginBy(Name|Spec)` methods could be confused, and same bugs may  appear in future: first instinct is "if err == nil, then main return value is usable". 
I think, `return nil, nil` is counter-intuitive and leads to additional check to `plugin == nil` in case of nil-plugin is unusable, like in `operationGenerator.GenerateDetachVolumeFunc`. This approach is only useful if we have lot of cases with usable nil-plugin .

I've checked all usages of `FindXXXXXXPluginBy(Name|Spec)`:

| Method name | with unusable `nil` plugin | with usable `nil` plugin | 
|-------------------|--------------------------------|------------------------------|
| FindAttachablePluginByName | 3 | 1 |
| FindAttachablePluginBySpec | 10 | 0 |
| FindDeviceMountablePluginBySpec | 2 | 0 |
| FindDeviceMountablePluginByName | 1 | 0 |
| FindExpandablePluginBySpec | 4 | 0 |
| FindExpandablePluginByName | 0 | 0 |
| FindMapperPluginBySpec | 1 | 0 |
| FindMapperPluginByName | 2 | 0 |
| FindFSResizablePluginBySpec | 0 | 1 |
| FindFSResizablePluginByName | 0 | 0 |
|**total** | **23** | **2** |

So, current approach is useful in only **2** cases in k8s code, and leads to surplus checks and potential nill pointer dereference in **23** cases.
I think, it's better to rewrite this methods to return nil error only with usable (non-nil) plugin, and add special error check (smth like [os.IsNotExist(err error) bool](https://golang.org/pkg/os/#IsNotExist)):  it's better to use additional checks in rare cases, not in regular.

liggitt commented:
if you want to pick this fix back to prior releases, doing the minimal fix first would be better.

I'd be supportive of a follow-up that changes the return values to make it clearer that there is a non-error case that does not return a usable plugin. As evidenced by the bug this is fixing, that is not always clear.

mshaverdo commented:
@liggitt @gnufied @saad-ali
It seems it's better to fix this bug by one-line commit checking `attachableVolumePlugin == nil`, and rewrite `FindXXXXXXPluginBy(Name|Spec)` functions in separate PR

mshaverdo committed:
Fix nil pointer dereference panic in attachDetachController

add check `attachableVolumePlugin == nil` to operationGenerator.GenerateDetachVolumeFunc()

mshaverdo commented:
@gnufied @saad-ali
I've changed PR to just fix panic, without changing signatures

mshaverdo commented:
/retest

gnufied commented:
Can you add a release note about fixing possible panic when plugin is not found?

mshaverdo commented:
@gnufied 
done

gnufied commented:
/lgtm
/approve

mshaverdo commented:
/kind bug

mshaverdo commented:
@gnufied
Could you please help? This PR stuck in "Merging is blocked" somehow. What should i do to merge it?

gnufied commented:
@mshaverdo it does not belong to 1.13 milestone, which is fine I think. A soon as master opens for general PR merges this will get merged. We will pick this for 1.13.1 and cherry-pick.

saad-ali commented:
Now that this is merged. Let's get it cherry picked for 1.13.1
